android-components/components/browser/menu2/src/main/java/mozilla/components/browser/menu2/ext/MenuPositioningData.kt

```diff
@@ -0,0 +1,269 @@
+import androidx.recyclerview.widget.RecyclerView
+import mozilla.components.browser.menu2.R
+import mozilla.components.concept.menu.MenuStyle
+import mozilla.components.concept.menu.Orientation
+import mozilla.components.support.base.log.Log
+import mozilla.components.support.ktx.android.view.isRTL
+import kotlin.math.roundToInt
+
+const val HALF_MENU_ITEM = 0.5
+
+@Suppress("ComplexMethod")
+internal fun inferMenuPositioningData(
+    containerView: View,
+    anchor: View,
+    style: MenuStyle? = null,
+    orientation: Orientation?,
+): MenuPositioningData? {
+    // Measure the menu allowing it to expand entirely.
+    val spec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
+    containerView.measure(spec, spec)
+
+    val recyclerView = containerView.findViewById<RecyclerView>(R.id.mozac_browser_menu_recyclerView)
+    val recyclerViewAdapter = recyclerView.adapter ?: run {
+        Log.log(
+            priority = Log.Priority.ERROR,
+            tag = "mozac-popupwindow",
+            message = "The browser menu recyclerview has no adapter set.",
+        )
+        return null
+    }
+
+    var containerViewHeight = containerView.measuredHeight
+    val containerViewWidth = containerView.measuredWidth
+    val recyclerViewHeight = recyclerView.measuredHeight
+    val recyclerViewItemCount = recyclerViewAdapter.itemCount
+
+    val hasViewInitializedCorrectly =
+        containerViewHeight > 0 && containerViewWidth > 0 && recyclerViewHeight > 0 && recyclerViewItemCount > 0
+    if (!hasViewInitializedCorrectly) {
+        Log.log(
+            priority = Log.Priority.ERROR,
+            tag = "mozac-popupwindow",
+            message = "containerViewHeight: $containerViewHeight must be more than 0," +
+                "recyclerViewHeight: $recyclerViewHeight must be more than 0 &" +
+                "recyclerViewItemCount: $recyclerViewItemCount must be more than 0.",
+        )
+        return null
+    }
+
+    // Elevation creates some padding between the menu and its container, so that the start corner
+    // of the menu doesn't match the corner of the anchor view. If the user wants the menu to hide
+    // the anchor completely, we have to adjust the position of the menu to compensate for the inner
+    // padding of the menu container.
+    val menuHorizontalPadding = containerViewWidth - recyclerView.measuredWidth
+    val menuVerticalPadding = containerViewHeight - recyclerView.measuredHeight
+    var horizontalOffset = style?.horizontalOffset ?: 0
+    var verticalOffset = style?.verticalOffset ?: 0
+
+    if (style?.completelyOverlap == true) {
+        horizontalOffset -= menuHorizontalPadding / 2
+        verticalOffset -= menuVerticalPadding / 2
+    }
+
+    // The menu height might be adjusted: if there is not enough space to show all the items,
+    // it will crop the last visible item in half, to give the user a hint that it is scrollable.
+    containerViewHeight = calculateContainerHeight(
```

Comment:
I was not sure if I should change the old `calculatedContainerHeight` but the problem with it – the majority of the time the function would just return the same height. It doesn't technically calculate the new height, it adjusted in case there is not enough space to show the full menu.